specifyweb/specify/scoping.py

```diff
@@ -1,96 +1,134 @@
         if institution.isaccessionsglobal:
-            return INSTITUTION_SCOPE, institution.id
+            return ScopeType.INSTITUTION, institution
         else:
             return self._simple_division_scope()
 
-    def agent(self): return self._simple_division_scope()
-
-    def borrow(self): return self._simple_collection_scope()
-
-    def collectingevent(self): return self._simple_discipline_scope()
-
-    def collectionobject(self): return self._simple_collection_scope()
-
-    def conservdescription(self): return self._simple_division_scope()
-
     def conservevent(self): return Scoping(self.obj.conservdescription)()
 
-    def dnasequence(self): return self._simple_collection_scope()
-
-    def dnasequencing(self): return self._simple_collection_scope()
-
-    def exchangein(self): return self._simple_division_scope()
-
-    def exchangeout(self): return self._simple_division_scope()
-
-    def fieldnotebook(self): return self._simple_discipline_scope()
-
     def fieldnotebookpage(self): return Scoping(self.obj.pageset)()
 
     def fieldnotebookpageset(self): return Scoping(self.obj.fieldnotebook)()
 
-    def gift(self): return self._simple_discipline_scope()
-
-    def loan(self): return self._simple_discipline_scope()
+    def gift(self):
+        if has_related(self.obj, 'discipline'):
+            return self._simple_discipline_scope()
 
-    def locality(self): return self._simple_discipline_scope()
+    def loan(self):
+        if has_related(self.obj, 'discipline'):
+            return self._simple_discipline_scope()
 
     def permit(self):
-        return INSTITUTION_SCOPE, self.obj.institution_id
-
-    def preparation(self): return self._simple_collection_scope()
+        if has_related(self.obj, 'institution'):
+            return ScopeType.INSTITUTION, self.obj.institution
 
     def referencework(self):
-        institution = models.Institution.objects.get()
-        return INSTITUTION_SCOPE, institution.id
-
-    def repositoryagreement(self): return self._simple_division_scope()
+        if has_related(self.obj, 'institution'):
+            return ScopeType.INSTITUTION, self.obj.institution
 
     def taxon(self):
-        return DISCIPLINE_SCOPE, self.obj.definition.discipline.id
+        return ScopeType.DISCIPLINE, self.obj.definition.discipline
 
 #############################################################################
 
-    def _simple_discipline_scope(self):
-        return DISCIPLINE_SCOPE, self.obj.discipline_id
-
-    def _simple_division_scope(self):
-        return DIVISION_SCOPE, self.obj.division_id
+    def _simple_discipline_scope(self) -> Tuple[int, Model]:
+        return ScopeType.DISCIPLINE, self.obj.discipline
+
+    def _simple_division_scope(self) -> Tuple[int, Model]:
+        return ScopeType.DIVISION, self.obj.division
+
+    def _simple_collection_scope(self) -> Tuple[int, Model]:
+        if hasattr(self.obj, "collectionmemberid"):
+            try:
+                """
```

Comment:
same as above

Reply:
Fixed in https://github.com/specify/specify7/pull/4215/commits/43aee5b97e52edca55b58e0cb2b76773754b510f